packages/@aws-cdk/aws-rds/lib/serverless-cluster.ts

```diff
@@ -299,6 +308,21 @@ abstract class ServerlessClusterBase extends Resource implements IServerlessClus
     });
   }
 
+  /**
+   * Grant the given identity to access to the Data API
+   *
+   * @param grantee The principal to grant access to
+   */
+  public grantDataApi(grantee: iam.IGrantable): iam.Grant {
+    const ret = iam.Grant.addToPrincipal({
+      grantee,
+      actions: DATA_API_ACTIONS,
+      resourceArns: [this.clusterArn],
```

Comment:
According to the docs, the resource on this policy is `'*'`. I'm not sure how limiting to the cluster ARN will work, especially since the policy includes actions across the rds-data, secretsmanager, and tag namespaces. My gut is that this would break the tag/secretsmanager interactions.

It's possible the policy could be split into two: the rds-data actions linked to the cluster ARN, and the other actions on '*', but that would need to be tested.

Reply:
The policies are a bit messy. I contacted the Data API team a while ago to clarify their use but didn't get a response. I'm using a lambda for 2 months using the Data API, with a policy limited to the ARN of the cluster, with no issues so far™.

My (previous) setup:

```ts
lambda.addToRolePolicy(
  new iam.PolicyStatement({
    actions: 'rds-data:*',
    resources: [clusterArn],
  })
);
lambda.grantRead(apiLambda);
```

<details>
<summary>Synthesized policy</summary>

```json
{
  "Action": "rds-data:*",
  "Effect": "Allow",
  "Resource": {
    "Fn::Join": [
      "",
      [
        "arn:aws:rds:eu-central-1:575357153522:cluster:",
        {
          "Fn::ImportValue": "Staging-Db:StagingDbExportsOutputRefClusterEB0386A72F7B91A7"
        }
      ]
    ]
  }
},
{
  "Action": [
    "secretsmanager:GetSecretValue",
    "secretsmanager:DescribeSecret"
  ],
  "Effect": "Allow",
  "Resource": {
    "Fn::ImportValue": "Staging-Db:StagingDbExportsOutputRefDbSecret685A0FA55952A3B0"
  }
}
```
</details>

This PR (to get closer to what is included in the [official docs](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/data-api.html#data-api.access)):

```ts
cluster.grantDataApi(lambda)
lambda.grantRead(apiLambda);
```

<details>
<summary>Synthesized policy</summary>

```json
{
  "Action": [
    "secretsmanager:CreateSecret",
    "secretsmanager:ListSecrets",
    "secretsmanager:GetRandomPassword",
    "tag:GetResources",
    "rds-data:BatchExecuteStatement",
    "rds-data:BeginTransaction",
    "rds-data:CommitTransaction",
    "rds-data:ExecuteStatement",
    "rds-data:RollbackTransaction"
  ],
  "Effect": "Allow",
  "Resource": {
    "Fn::Join": [
      "",
      [
        "arn:aws:rds:eu-central-1:575357153522:cluster:",
        {
          "Fn::ImportValue": "Staging-Db:StagingDbExportsOutputRefClusterEB0386A72F7B91A7"
        }
      ]
    ]
  }
},
{
  "Action": [
    "secretsmanager:GetSecretValue",
    "secretsmanager:DescribeSecret"
  ],
  "Effect": "Allow",
  "Resource": {
    "Fn::ImportValue": "Staging-Db:StagingDbExportsOutputRefDbSecret685A0FA55952A3B0"
  }
}
```
</details>

[Official docs](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/data-api.html#data-api.access):

```ts
cluster.grantDataApi(lambda)
lambda.grantRead(apiLambda);
```

<details>
<summary>Synthesized policy</summary>

```json
{
  "Sid": "SecretsManagerDbCredentialsAccess",
  "Effect": "Allow",
  "Action": [
      "secretsmanager:GetSecretValue",
      "secretsmanager:PutResourcePolicy",
      "secretsmanager:PutSecretValue",
      "secretsmanager:DeleteSecret",
      "secretsmanager:DescribeSecret",
      "secretsmanager:TagResource"
  ],
  "Resource": "arn:aws:secretsmanager:*:*:secret:rds-db-credentials/*"
},
{
  "Sid": "RDSDataServiceAccess",
  "Effect": "Allow",
  "Action": [
      "secretsmanager:CreateSecret",
      "secretsmanager:ListSecrets",
      "secretsmanager:GetRandomPassword",
      "tag:GetResources",
      "rds-data:BatchExecuteStatement",
      "rds-data:BeginTransaction",
      "rds-data:CommitTransaction",
      "rds-data:ExecuteStatement",
      "rds-data:RollbackTransaction"
  ],
  "Resource": "*"
}
```
</details>

Not that the "secret" statement is also very different. The official docs grant get/put/delete, while this PR grants get only. The grants in the official docs look very permissive ... maybe too permissive?

Reply:
Your original (previous) setup does look somewhat more ideal, not knowing the impact of dropping the extra Secrets Manager and tagging permissions. I also found one other source (https://www.jeremydaly.com/aurora-serverless-data-api-a-first-look/) where someone marked most of the other permissions as "unnecessary" for their jobs.

My preference is to start with minimal permissions we've validated are necessary, and add more later as new use cases are discovered; it's much easier for us to add permissions than remove them! 

To that end, perhaps we should only included the "baseline" permissions for the rds-data API (for now):

<details>
<summary>Example policy</summary>

```json
{
  "Action": [
    "rds-data:BatchExecuteStatement",
    "rds-data:BeginTransaction",
    "rds-data:CommitTransaction",
    "rds-data:ExecuteStatement",
    "rds-data:RollbackTransaction"
  ],
  "Effect": "Allow",
  "Resource": "arn:aws:rds:eu-central-1:575357153522:cluster:MyCluster",
},
{
  "Action": [
    "secretsmanager:GetSecretValue",
    "secretsmanager:DescribeSecret"
  ],
  "Effect": "Allow",
  "Resource": "arn:aws:secretsmanager:....",
}
```
</details>

Tagging @shivlaks  and @skinny85  for their thoughts as well.

Reply:
I'm wondering now if the extra SecretsManager-related permissions are for creating users and updating user passwords? Regardless, still happy to stick with the minimum for now and add as necessary.

Reply:
I'm also in favour of minimal permissions that we can validate.

At a first glance it does **not** look like RDS Data API supports [resource-level permissions, policies](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_aws-services-that-work-with-iam.html#database_svcs), or tag based authorization. It's worth validating the documentation as well.

Is a general `grant` method also possible where users would supply a principal and list of actions?
That serves as an escape hatch for added APIs or grant methods that are too restrictive for some use cases.

Reply:
Should that general `grant` method only apply to the Data API (**without** resource-level permissions)? Because there could also be a `grantConnect` method for the cluster's `connections` (**with** resource-level permissions) as for `IDatabaseInstance`:

https://github.com/aws/aws-cdk/blob/09da33a92ecca07b7a275579c078b7a626424b94/packages/%40aws-cdk/aws-rds/lib/instance.ts#L66-L69

So, should we go with a a general `grantDataApi` (the general one) and a `grantDataApiExecute` (the "default" one)?